### ISSUE TITLE
Nested `Date` type attributes are saved differently at different levels

### ISSUE THREAD

**Issue by hariprasadiit** [parse-community/parse-server#7575]
### New Issue Checklist

- [x] I am not disclosing a [vulnerability](https://github.com/parse-community/parse-server/blob/master/SECURITY.md).
- [x] I am not just asking a [question](https://github.com/parse-community/.github/blob/master/SUPPORT.md).
- [x] I have searched through [existing issues](https://github.com/parse-community/parse-server/issues?q=is%3Aissue).
- [x] I can reproduce the issue with the [latest version of Parse Server](https://github.com/parse-community/parse-server/releases). 

### Issue Description

If we try to create an object with below code

```js
const myObj = new Parse.Object('Company3');
await myObj.save({
  prop1: 'test1',
  prop2: {
    test: {
      date: new Date()
    }
  },
  prop3: {
    date: new Date()
  }
}, {useMasterKey: true});
```

This is how the saved object in mongo DB looks like

```json
{
  "_id": "5DTIqZcZv9fhbHu0qTZq",
  "prop2": {
    "test": {
      "date": {
        "__type": "Date",
        "iso": "2021-09-14T14:50:37.303Z"
      }
    }
  },
  "prop3": {
    "date": {
      "$date": "2021-09-14T14:50:37.303Z"
    }
  },
  "prop1": "test1",
  "_created_at": {
    "$date": "2021-09-14T14:50:38.108Z"
  },
  "_updated_at": {
    "$date": "2021-09-14T14:50:38.108Z"
  }
}
```

`prop2.test.date` is saved as JSON object instead of `$date` object while `prop3.date` is properly saved.
This breaks date queries

### Steps to reproduce
1. Create new object with below code

```js
const myObj = new Parse.Object('Company3');
await myObj.save({
  prop1: 'test1',
  prop2: {
    test: {
      date: new Date()
    }
  },
  prop3: {
    date: new Date()
  }
}, {useMasterKey: true});
```
2. Observe different date types at different levels in mongoDB Compass

### Actual Outcome
```json
{
  "_id": "5DTIqZcZv9fhbHu0qTZq",
  "prop2": {
    "test": {
      "date": {
        "__type": "Date",
        "iso": "2021-09-14T14:50:37.303Z"
      }
    }
  },
  "prop3": {
    "date": {
      "$date": "2021-09-14T14:50:37.303Z"
    }
  },
  "prop1": "test1",
  "_created_at": {
    "$date": "2021-09-14T14:50:38.108Z"
  },
  "_updated_at": {
    "$date": "2021-09-14T14:50:38.108Z"
  }
}
```

### Expected Outcome
```json
{
  "_id": "5DTIqZcZv9fhbHu0qTZq",
  "prop2": {
    "test": {
      "date": {
        "$date": "2021-09-14T14:50:37.303Z"
      }
    }
  },
  "prop3": {
    "date": {
      "$date": "2021-09-14T14:50:37.303Z"
    }
  },
  "prop1": "test1",
  "_created_at": {
    "$date": "2021-09-14T14:50:38.108Z"
  },
  "_updated_at": {
    "$date": "2021-09-14T14:50:38.108Z"
  }
}
```

### Environment

Server
- Parse Server version: `4.10.3`
- Operating system: `Mac OS`
- Local or remote host (AWS, Azure, Google Cloud, Heroku, Digital Ocean, etc): `Local`

Database
- System (MongoDB or Postgres): `MongoDB`
- Database version: `4.4.4`
- Local or remote host (MongoDB Atlas, mLab, AWS, Azure, Google Cloud, etc): `Local`

Client
- SDK (iOS, Android, JavaScript, PHP, Unity, etc): `Javascript`
- SDK version: `3.3.0`

Old issue related to this #6840

**Comment by mtrezza** [comment]
I was able to reproduce the issue. Would you want to open a PR with a failing test?

**Comment by hariprasadiit** [comment]
@mtrezza Please check the PR

**Comment by cbaker6** [comment]
I suspect this is more of a JS SDK issue. According to the [REST documentation](https://docs.parseplatform.org/rest/guide/#data-types), Parse treats `Date` as a special data type:

> The Date type contains a field iso which contains a UTC timestamp stored in ISO 8601 format with millisecond precision: YYYY-MM-DDTHH:MM:SS.MMMZ.
```json
{
  "__type": "Date",
  "iso": "2022-01-01T12:23:45.678Z"
}
```
 
So the outcome of `prop2` looks correct IMO and the others are wrong:

```json
"prop2": {
    "test": {
      "date": {
        "__type": "Date",
        "iso": "2021-09-14T14:50:37.303Z"
      }
    }
  }
```

My explanation https://github.com/parse-community/parse-server/issues/8001#issuecomment-1133631026 may provide some more context on what the JS SDK should be doing. From your current post it seems the JS SDK is partially encoding date correctly, but not handling date correctly on top level dates (i.e. `prop3`)

I suspect the issue is here in the JS SDK: https://github.com/parse-community/Parse-SDK-JS/blob/e2bd6be145f55fab322633a8f144be34e7fcb202/src/encode.js#L62-L67. If you add an if statement to check if the current top level object is a date and use a similar block transforming the date to the iso format, it should fix this issue.

**Comment by parseplatformorg** [comment]
🎉 This change has been released in version [6.0.0-alpha.15](https://github.com/parse-community/parse-server/releases/tag/6.0.0-alpha.15)

**Comment by parseplatformorg** [comment]
🎉 This change has been released in version [6.0.0-beta.1](https://github.com/parse-community/parse-server/releases/tag/6.0.0-beta.1)

**Comment by parseplatformorg** [comment]
🎉 This change has been released in version [6.0.0](https://github.com/parse-community/parse-server/releases/tag/6.0.0)
